### DISCUSSION TITLE
How to reset sub-app data after each time step of main app in a Multi app simulation

### DISCUSSION THREAD

**Discussion by sSajjad90** [idaholab/moose#23945] [Q&A General] [answered]
Hi,
I'm attempting to run a multi-app simulation that includes a full sub-app solution at each time step of the main app. My issue is figuring out how to reset sub-app data after each time step of the main app.
I noticed that the sub-model resumes the analysis from the previous time step of the main model, which contradicts my demand for a fresh start after each time step.
Moreover, when I attempted to use `reset_ apps` and `rest_time,` I received "Segmentation fault (core dumped)".
How can I reset sub-app data after each time step of the main app?

And this is the `Multi app`
```
`[MultiApps]
  [full_solve]
    type = FullSolveMultiApp
    input_files = Nucleation.i
    execute_on = TIMESTEP_BEGIN
    positions =  '0 0 0'
    no_backup_and_restore = true
    reset_apps = '0'
    reset_time = '5.0'

  [../]
[]`
```

**Comment by GiudGiud** [comment]
Hello

You dont need the reset time / app parameter, just set the booleans to get a reset and you ll get a reset every time.

I think there's another one, not just no_backup_and_restore

Guillaume

**Comment by sSajjad90** [comment]
I'm not sure what you mean by "set the booleans."
I eliminated the reset time / app parameter, as well as the no backup and restore option. Yet, the sub-app continues to use the previous iteration value, as follows:
```
[MultiApps]
  [full_solve]
    type = FullSolveMultiApp
    input_files = Nucleation.i
    execute_on = TIMESTEP_BEGIN
  [../]
[]
```
![multiapp2](https://user-images.githubusercontent.com/78385504/229623447-9d24efed-1d01-4892-a397-713bbe59c9a9.png)

**Reply by GiudGiud** [reply]
set keep_solution_during_restore = false see if it helps
no_backup_and_restore = true should also be there

**Reply by sSajjad90** [reply]
It had no effect.

**Comment by sSajjad90** [comment]
Do you think the source code ( `MultiApp.C `)  is working properly? 
Could you please take a kind look at the relevant source code  `MultiApp.C ` at [https://github.com/idaholab/moose/blob/next/framework/src/multiapps/MultiApp.C](url), line 638 to 713 to make sure it is working correctly for the `TIMESTEP_BEGIN`

Moreover, there is a topic (#18549) with a similar problem.

Thanks for your support,
Sajjad

**Reply by GiudGiud** [reply]
I ll try to have a look today

**Reply by sSajjad90** [reply]
Many Thanks,

**Reply by GiudGiud** [reply]
ok so resetting (reset_time/app) does not work with FullSolveMultiApp.

I ll create an issue.

**Comment by GiudGiud** [comment, marked as answer]
ok so we just do not have a clean mechanism to reset a FullSolveMultiApp at this point

reset is not working, and the backup/restore things are meant for fixed point iterations, not time step to time step. 

HOWEVER, what you can do:
- for auxvariables, use an auxkernel on initial to reset them to their initial condition:
```
[AuxKernels]
  [zero_out]
    type = ParsedAux
    variable = u
    expression = 0 # could be the initial condition instead
    execute_on = TIMESTEP_BEGIN # initial would be better but that does not work. Keep in mind this will run a LOT
  []
[]
```
- for main variables, use a transfer from the parent app
```
[AuxVariables]
  [zero]
  []
[]
```

```
[Transfers]
  [reset]
    type = MultiAppNearestNodeTransfer
    to_multi_app = sub
    source_variable = 'zero'  # this could be the initial condition for that variable instead
    variable = v # main variable to reset
  []

```

**Reply by GiudGiud** [reply]
This is the issue for this
https://github.com/idaholab/moose/issues/9539

**Comment by Bearunner** [comment]
Hello Guillaume @GiudGiud,

thanks for your answer 'For main variables, use a transfer from the parent app'. It helps me solve the zero setting of one main variable.

But I have another sub-app main variable (i.e., pore pressure) reset to hydrostatic values (e.g., p=p0+rho*g*h increasing with depth)  at a specific time.

It works well if their mesh are same or at least has the same vertical extension scale. But in my case, the sub-app extent is from 0 to 500m depth. The main app only extends from 400m to 500m depth.

![image](https://github.com/idaholab/moose/assets/126703042/d6595a68-bb62-40e3-a636-799f1d39b366)

Only the p in their intersecting /overlapping areas (400-500 depth) have the same value to reference hydrostatic values. 

Could you please help me with this? Thanks a lot.

B

**Reply by GiudGiud** [reply]
Hello

So you want the field in the subapp to be reset even though its mesh extends beyond the one in the main app?

I think you could do this with siblings transfer.
Duplicate the multiapp for the subapp. Except one is inactive (just execute it on initial) and does not do anything, just act as a source for fields.

```
[MultiApps]
  [true_subapp]
   ...
  []
  [inactive_one]
    type = ...
    execute_on = INITIAL
    cli_args = 'Problem/solve=false'
    ...
  []
[]
```

then transfer from the inactive one to the real one
```
[Transfers]
  [sibling_pore_p]
    type = MultiAppGeneralFieldShapeEvaluationTransfer
    from_multiapp = inactive_one
    to_multiapp = true_subapp
    from_variable = pore_pressure
    to_variable = pore_pressure
  []
[]
```

**Reply by Bearunner** [reply]
Thanks for your answer. It works well for my problem. Have a nice weekend!
